User Epic

alrzi opened the pull request:
Проект написан на MVVM) Мне разрешили писать на нем, хотя ребята пишут на MVP. 
[Statistic.md](https://github.com/miamib34ch/YP-CP-FakeNFT/files/12386248/Statistic.md)

alrzi committed:
#39 - Добавил экран со списком юзеров

alrzi committed:
#38 - Добавил возможность сортировки

alrzi committed:
#36 - Добавил экран с информацией о пользователе

alrzi committed:
#35 - Добавил коллекцию с нфт

alrzi committed:
#35 - Переместил файлы

alrzi committed:
#35 - Переместил фалы

alrzi committed:
Merge branch 'base' into user

# Conflicts:
#	FakeNFT.xcodeproj/project.pbxproj

alrzi committed:
#35 - Добавил в коллекцию NFT показ лайков и в корзине ли NFT

alrzi committed:
#35 - Удалил ненужные проверки

alrzi committed:
№35 - Удалил ошибку

alrzi committed:
#35 - Убрал рефреш контрол

miamib34ch commented:
Привет! Посмотрел твою работу и нашёл моменты, которые нужно исправить:  

- Добавить в пулл реквест комментарий для ревьювера Практикума
- В таббаре в иос < 15 у твоего раздела нет картинки, проект поддерживается с 13ой иос
- Пагинация отсутствует при выгрузке общего списка пользователей
- В лог выводятся непонятные цифры, нужно почистить принты
- Сортировка по количеству работает неправильно (скинул скрин бага в беседу)
- Сортировка по умолчанию неправильная - должна быть по рейтингу

Также заметил, что сервис сортировки не общий и лайки и корзина не работают, но может ты так и хотел who knows

alrzi commented:
Какой комментарий я должен добавитьдля ревьюера?

miamib34ch commented:
> Какой комментарий я должен добавитьдля ревьюера?

В вашем Pull Request приложена ссылка на таск трекер. В доске должна быть информация о вашем прогрессе:

- выбранная командой архитектура и способ верстки
- декомпозиция эпика (ссылка на таск-трекер)
- декомпозиция эпика (приложенный текстовый документ .md)
- первоначальная оценка времени выполнения задач
- реальное время, потраченное на решении каждой задачи

Обратите внимание на последний пункт – потраченное время. Это важно для улучшения ваших навыков декомпозиции. Сравнивая первоначальную оценку с реальным временем, вы можете отследить, насколько реалистично оцениваете свои силы и полностью ли предусмотрели шаги в задаче. Зная это, в следующий раз вы сможете дать более точный прогноз.

alrzi committed:
#35 - Чуть исправил код

alrzi committed:
Merge branch 'base' into user

# Conflicts:
#	FakeNFT.xcodeproj/project.pbxproj

alrzi committed:
#40 - Добавлен новый с сайтом пользователя, чуть исправлен код

alrzi committed:
Merge branch 'base' of https://github.com/miamib34ch/YP-CP-FakeNFT into user

alrzi committed:
#40 - Добавил вью модель для WebView

alrzi committed:
#40 - Убрал CurrentValueSubject так как это класс и value изменяемое

LizaIlinykh commented on FakeNFT/Epics/Statistic/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    func configureUI(with nfts: [NFTCellViewModel]) {
+        if !nfts.isEmpty {
```
мне не очень нравится , что вьюха сама решает, что ей показать в зависимости от того пустые ли нфт ей приходят. если показ таблицы или заглушки будет зависеть в дальнейшем еще и от других факторов(или например мы захотим показать другую заглушку) эта бизнес логика будет во вьюхе разрастаться. я бы рекомендовала не посылать событие  nftViewModels, если массив с нфт пустой, а послать другое и показать заглушку(и передать как раз модель заглушки в ее собственном событии). в целом это по желанию правка, сейчас бизнес логики не так много, но на реальном проекте это будет не так =)

LizaIlinykh commented on FakeNFT/Epics/Statistic/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,220 @@
+                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+                    self.updateSnapshot(with: nfts)
+                    self.collectionView.refreshControl?.endRefreshing()
+                }
```
а тут зачем ожидание в секунду?

alrzi replied:
анимация выглядит более плавной и в целом более приятной, если убрать то будет очень быстро возвращаться после пула

LizaIlinykh replied:
это не анимация, ты просто откладываешь выполнение на 1 секунду

alrzi replied:
я понимаю) но если это не сделать тогда после оттягивания (pull to refresh) она очень быстро обновится и если интернет хороший тогда мигом улетит наверх) я как пользователь не хотел бы видеть такое поведение в приложении так как это выглядит не естественно, дерганно) лучше немного подождать ( секунда это не долго) и дать понять что данные действительно обновились)

LizaIlinykh commented on FakeNFT/Epics/Statistic/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,220 @@
+                self.collectionView.refreshControl?.endRefreshing()
+            }
+        } else {
+            // TODO: - Добавить заглушку если массив пустой
```
надо добавить =)

LizaIlinykh commented on FakeNFT/Epics/Statistic/NFTCollection/NFTCollectionViewModel.swift:
```diff
@@ -0,0 +1,118 @@
+        var nfts = userNfts
+
+        for (index, nft) in nfts.enumerated() {
+            let id = Int(nft.id) ?? 0
```
я бы сделала здесь guard и если не получается сделать число выбрасывала бы ошибку
быстрее так дебажить

alrzi replied:
Ты бы сделала функцию которая выбрасывает ошибки через throw?

LizaIlinykh replied:
я так подумала я бы просто assertionFailure здесь вызвала если не получится число сделать. прям ошибку выкидывать лишнее

LizaIlinykh commented on FakeNFT/Epics/Statistic/Services/UserService/UserService.swift:
```diff
@@ -0,0 +1,42 @@
+            return
+        }
+
+        let task = defaultNetworkClient.send(request: userResult, type: [UserResult].self) { result in
```
слабый селф, замыкание сбегающее, может быть retain cycle

LizaIlinykh commented on FakeNFT/Epics/Statistic/UserCard/SectionsCellsView/CollectionCellView.swift:
```diff
@@ -0,0 +1,75 @@
+final class CollectionCellView: UICollectionViewCell {
+    func configure(with viewModel: NFTCollectionCellViewModel) {
+        nftNumberInCollection.text = viewModel.formattedNftsNumber
+    }
```
вот пример класса с нарушенной иерархией
оставлю на платформе ссылку

LizaIlinykh commented on FakeNFT/Epics/Statistic/WebView/WebViewViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+}
+
+final class WebViewModelImpl: WebViewModel {
+    let url = URL(string: "https://practicum.yandex.ru")
```
не нужно хардкодить, давай с инициализатора занесем готовую модель url request и будем отдавать ее вью для показа

LizaIlinykh replied:
как считаешь зачем вью модели урл если она им не пользуется?)

alrzi replied:
я смогу эту вью модель использовать в другом контроллере с webView и с другой url если запихать с инита) А ты как считаешь ? На самом деле я наверно это делал не очень осознано, просто подумал что это к вью слою не относится.

LizaIlinykh replied:
все верно к вью слою не относится
просто логика мввм нарушена, поэтому выглядит так странно
если в вью есть слово урл реквест значит что то не так)
да и вообще сама вьюха не должна данные тянуть, ей должна их отдать вью модель
и когда вью модель САМА (она же главная в модуле) решит , что пора загрузить вебвью , она должна вьюху об этом попросить передав ей в биндинге урл реквест

в общем слишком самостоятельная вьюшка = не оч хорошо

LizaIlinykh commented on FakeNFT/TabBar/TabBar.swift:
```diff
@@ -0,0 +1,97 @@
+    init() {
+        super.init(nibName: nil, bundle: nil)
+
+        let statisticMediumConfigForImage = UIImage.SymbolConfiguration(scale: .medium)
```
надо это все в отдельный метод и таббар вы оставили без слоя вью модели - надо бы доделать для красоты

alrzi replied:
Мне разрешили писать на MVVM хотя ребята пишут на MVP. Поэтому таб бар трогать не буду)

LizaIlinykh commented on FakeNFT/TabBar/TabBar.swift:
```diff
@@ -0,0 +1,97 @@
+
+        let viewModel = UserListViewModelImpl(userStatisticService: userService)
+
+        return UserListViewController(viewModel: viewModel)
```
не очень нравится что таббар занимается сборкой модулей, лучше вынести в фабрику

alrzi replied:
да) я знаю что так делать не надо. Но оставлю так.

LizaIlinykh commented on FakeNFT/TabBar/TabBar.swift:
```diff
@@ -0,0 +1,97 @@
+
+        return UserListViewController(viewModel: viewModel)
+    }
+}
```
в целом, работа очень классная
круто, что использовал diffable data source, combine, реализовал pull to refresh
есть только мини - замечания, думаю их несложно тебе будет поправить =)

alrzi committed:
Поправил замечания

1) - В списке пользователей справа отображается количество нфт, а не рейтинг
2) - Добавить заглушку, если коллекция нфт пустая
3) - Вёрстка. Auto Layout (поправил баг в ячейке UserCell)
4) - Чистота кода ( добавил MARK - )

alrzi committed:
Исправил замечания

alrzi committed:
Добавил throttle для кнопок

alrzi committed:
Убрал принты

alrzi committed:
Merge remote-tracking branch 'origin/develop' into user

# Conflicts:
#	FakeNFT.xcodeproj/project.pbxproj
#	FakeNFT/Resources/Image.swift
